File: tools/query_dataset.py
"""
Tool: datagroom_query_dataset - Query dataset with structured filters (matches TS queryDataset.ts).
"""

import logging
from urllib.parse import quote

from schemas import Filter
from utils.authenticated_request import make_authenticated_request
from utils.error_handlers import format_error
from utils.formatters import format_markdown_table, format_query_summary

logger = logging.getLogger(__name__)

QUERY_DATASET_DESCRIPTION = """Query a Datagroom dataset with structured filters and return matching rows.

This tool allows you to filter, sort, and paginate through dataset results.

Args:
  - dataset_name (string, required): Name of the dataset to query
  - filters (array, optional): Array of filter objects with:
    - field: Field name to filter on
    - type: Filter type (eq, ne, gt, lt, gte, lte, in, nin, regex)
    - value: Filter value (array for 'in'/'nin', string for 'regex')
  - sort (object, optional): Sort configuration with:
    - field: Field name to sort by
    - direction: 'asc' or 'desc'
  - max_rows (number, optional, default: 100, max: 1000): Maximum rows to return
  - offset (number, optional, default: 0): Number of rows to skip (for pagination)
  - response_format (string, optional, default: 'markdown'): 'markdown' or 'json'

Returns:
  Object containing:
  - dataset_name: Name of the dataset
  - query_summary: Formatted summary of the query
  - total_matching: Total number of rows matching filters
  - rows_returned: Number of rows in this response
  - offset: Offset used
  - has_more: Whether more rows are available
  - next_offset: Offset for next page (if has_more is true)
  - data: Array of matching rows
  - warning: Warning message if results truncated

Examples:
  - Find transactions > $1000: filters=[{field: "amount", type: "gt", value: 1000}]
  - Get active users sorted by name: filters=[{field: "status", type: "eq", value: "active"}], sort={field: "name", direction: "asc"}
  - Paginate results: offset=100, max_rows=50"""


class QueryFilterInput(BaseModel):
    field: str
    type: str  # enum validated by Gateway
    value: object = None


class SortInput(BaseModel):
    field: str
    direction: str  # 'asc' | 'desc'


async def datagroom_query_dataset(
    dataset_name: str,
    filters: list[dict] | None = None,
    sort: dict | None = None,
    max_rows: int = 100,
    offset: int = 0,
    response_format: str = "markdown",
):
    """Query a dataset via Gateway with filters, sort, and pagination."""
    if not dataset_name or not dataset_name.strip():
        raise ValueError("Dataset name is required")
    if max_rows < 1 or max_rows > 1000:
        raise ValueError("max_rows must be between 1 and 1000")
    if offset < 0:
        raise ValueError("offset must be >= 0")
    filters = filters or []
    sorters = [sort] if sort else []
    page = offset // max_rows + 1 if max_rows else 1
    try:
        response = await make_authenticated_request(
            f"/ds/viewViaPost/{quote(dataset_name, safe='')}/default/mcp",
            "POST",
            {
                "filters": filters,
                "sorters": sorters,
                "page": page,
                "per_page": max_rows,
            },
        )
    except Exception as e:
        logger.exception("query_dataset failed")
        raise RuntimeError(f"Error querying dataset: {format_error(e)}") from e
    total = response.get("total") or 0
    data = response.get("data") or []
    rows_returned = len(data)
    has_more = offset + rows_returned < total
    filter_objs = [Filter(**f) for f in filters] if filters else []
    summary = format_query_summary(
        dataset_name, total, rows_returned, filter_objs, offset, has_more
    )
    data_table = format_markdown_table(data)
    text = f"{summary}\n\n{data_table}"
    from fastmcp.tools.tool import ToolResult
    return ToolResult(
        content=text,
        structured_content=response,
    )

File: src/tools/queryDataset.ts
/**
 * Tool: datagroom_query_dataset
 * Query dataset with structured filters
 */

import { Server } from '@modelcontextprotocol/sdk/server/index.js';
import { z } from 'zod';
import { MongoClient } from 'mongodb';
import { formatQuerySummary, formatMarkdownTable } from '../utils/formatters.js';
import { formatError } from '../utils/errorHandlers.js';
import { makeAuthenticatedRequest } from '../utils/authenticatedRequest.js';

const FilterSchema = z.object({
  field: z.string(),
  type: z.enum(['eq', 'ne', 'gt', 'lt', 'gte', 'lte', 'in', 'nin', 'regex']),
  value: z.any()
});

const SortSchema = z.object({
  field: z.string(),
  direction: z.enum(['asc', 'desc'])
}).optional();

const QueryDatasetInputSchema = z.object({
  dataset_name: z.string().min(1, 'Dataset name is required'),
  filters: z.array(FilterSchema).optional().default([]),
  sort: SortSchema.optional(),
  max_rows: z.number().int().min(1).max(1000).optional().default(100),
  offset: z.number().int().min(0).optional().default(0),
  response_format: z.enum(['markdown', 'json']).optional().default('markdown')
}).strict();

export function registerQueryDatasetTool(
  server: Server,
  client: MongoClient,
  toolHandlers: Map<string, (request: any) => Promise<any>>
) {
  toolHandlers.set('datagroom_query_dataset', async (request: any) => {
    try {
      const params = QueryDatasetInputSchema.parse(request.params.arguments);
      
      // Route through Gateway API (NOT direct MongoDB)
      // This ensures ACLs are enforced
      const response = await makeAuthenticatedRequest(
        `/api/dataset/${encodeURIComponent(params.dataset_name)}/query`,
        'POST',
        {
          filters: params.filters || [],
          sorters: params.sort ? [params.sort] : [],
          page: Math.floor((params.offset || 0) / (params.max_rows || 100)),
          per_page: params.max_rows || 100
        }
      );
      
      // Format response
      const summary = formatQuerySummary(
        params.dataset_name,
        response.total || 0,
        (response.data || []).length,
        params.filters || [],
        params.offset || 0,
        params.response_format || 'markdown'
      );
      
      const dataTable = formatMarkdownTable(response.data || []);
      
      return {
        content: [{
          type: 'text',
          text: `${summary}\n\n${dataTable}`
        }],
        structuredContent: response
      };
    } catch (error: any) {
      return {
        content: [{
          type: 'text',
          text: `Error querying dataset: ${formatError(error)}`
        }],
        isError: true
      };
    }
  });
}
